File: 剑指offer/61.扑克牌中的顺子.py
from typing import List


class Solution:
    def isStraight(self, nums: List[int]) -> bool:
        nums_a = []
        cnt = 0

        for num in nums:
            if num == 0:
                cnt += 1
            else:
                nums_a.append(num)

        nums_a.sort()

        for i in range(len(nums_a)):
            if i != 0:
                diff = nums_a[i] - nums_a[i - 1] - 1
                if diff < 0:
                    return False
                cnt -= diff
                if cnt < 0:
                    return False

        return True

File: 剑指offer/61.扑克牌中的顺子.cpp
class Solution {
public:
    bool isStraight(vector<int>& nums)
    {
        int num = 0;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] == 0) {
                num += 1;
            } else {
                if (i > 0 && nums[i - 1] != 0) {
                    int diff = nums[i] - nums[i - 1] - 1;
                    if (diff < 0) {
                        return false;
                    }
                    num -= diff;
                    if (num < 0) {
                        return false;
                    }
                }
            }
        }

        return true;
    }
};